core/src/sumeragi/main_loop.rs

```diff
@@ -463,123 +465,129 @@ fn handle_role_agnostic_messages<F>(
+                commit_block(sumeragi, state, block);
             }
             Message::BlockCommitted(BlockCommitted { hash, signatures }) => {
-                let block = if voting_block_option.is_some()
-                    && hash
-                        == voting_block_option
-                            .as_ref()
-                            .expect("Voting block is `Some`")
-                            .block
-                            .hash()
-                {
-                    let mut block = voting_block_option
-                        .take()
-                        .expect("Voting block is `Some`")
-                        .block;
-                    block.signatures.extend(signatures);
-                    block
-                } else {
-                    error!("The hash of the committed block does not match the hash of the block stored by the peer.");
-                    return;
-                };
+                if let Some(mut voted_block) = voting_block.take() {
+                    let voting_block_hash = voted_block.block.hash();
 
-                let block = match block.commit(&state.current_topology) {
-                    Ok(block) => block.into(),
-                    Err(err) => {
-                        warn!(?err);
-                        return;
-                    }
-                };
+                    if hash == voting_block_hash.transmute() {
+                        add_signatures::<true>(&mut voted_block, signatures.transmute());
 
-                commit_block(sumeragi, block, state);
+                        match voted_block.block.commit(&state.current_topology) {
+                            Ok(committed_block) => commit_block(sumeragi, state, committed_block),
+                            Err((_, err)) => {
+                                error!(%addr, %role, %hash, ?err, "Block failed to be committed")
+                            }
+                        };
+                    } else {
+                        error!(
+                            %addr, %role, committed_block_hash=%hash, %voting_block_hash,
+                            "The hash of the committed block does not match the hash of the block stored by the peer."
+                        );
+
+                        *voting_block = Some(voted_block);
+                    };
+                } else {
+                    error!(%addr, %role, %hash, "Peer missing voting block")
+                }
             }
-            other => *maybe_incoming_message = Some(other),
+            other => *incoming_message = Some(other),
         }
     }
 }
 
 #[allow(clippy::too_many_arguments)]
-fn compare_view_change_index_and_block_height_to_old<F>(
-    sumeragi: &SumeragiWithFault<F>,
+fn reset_state(
+    addr: &str,
     current_view_change_index: u64,
     old_view_change_index: &mut u64,
     current_latest_block_height: u64,
     old_latest_block_height: &mut u64,
```

Comment:
Should be a struct, no? 


There's something called negative cost abstractions, widely regarded in C++, when you can save lots of cycles by packing the data into a struct instead of passing each individually. It's an LLVM optimisation so it should work here too.

Reply:
I understand, if I continue with a refactor I will make this change in another PR. Considering the state of sumeragi code that I found I think this is much better now.